QueryEngine/ResultSetConversion.cpp

```diff
@@ -518,6 +526,65 @@ arrow::RecordBatch ResultSet::convertToArrow(const std::vector<std::string>& col
   return arrow::RecordBatch(schema, row_count, columns);
 }
 
+// WARN(ptaylor): users are responsible for detaching and removing shared memory segments, e.g.,
+//   int shmid = shmget(...);
+//   char *ipc_ptr = (char* shmat(shmid, ...));
+//   ...
+//   shmdt(ipc_ptr);
+//   shmctl(shmid, IPC_RMID, 0);
+std::tuple<std::shared_ptr<arrow::Buffer>, std::vector<char>, int64_t> ResultSet::getArrowCopy(
+    Data_Namespace::DataMgr* data_mgr,
+    const std::vector<std::string>& col_names) const {
+  auto arrow_copy = convertToArrow(col_names);
+  auto serialized_schema = serialize_arrow_schema(*arrow_copy.schema());
+  auto serialized_records = serialize_arrow_records(arrow_copy);
+  // Generate a new key for a shared memory segment. Keys to shared memory segments
+  // are OS global, so we need to try a new key if we encounter a collision. It seems
+  // incremental keygen would be deterministically worst-case. If we use a hash
+  // (like djb2) + nonce, we could still get collisions if multiple clients specify
+  // the same nonce, so using rand() in lieu of a better approach
+  // TODO(ptaylor): Is this common? Are these assumptions true?
+  key_t key = rand();
+  int shmsz = (int) serialized_records->size();
+  int shmid;
+  // IPC_CREAT - indicates we want to create a new segment for this key if it doesn't exist
+  // IPC_EXCL - ensures failure if a segment already exists for this key
+  while ((shmid = shmget(key, shmsz, IPC_CREAT | IPC_EXCL | 0666)) < 0) {
```

Comment:
I'm no expert in IPC, but from what @lmeyerov and I could tell, it seemed a file descriptor based solution could potentially run up against system limits on the number of simultaneous open FDs per process. Additionally, the node lib for attaching to shared memory segments is more active, and seems to reliably `shmctrl` on GC, even if the process crashes. Though I'm happy to defer to the expertise of the team here.